src/lowering.rs

```diff
@@ -251,6 +254,66 @@ impl CrateReader {
         }
     }
 
+    fn read_borrows<L: AnalysisLoader>(&self,
+            per_fn_borrows: Vec<raw::BorrowData>,
+            analysis: &mut PerCrateAnalysis,
+            project_analysis: &AnalysisHost<L>) {
+        for b in per_fn_borrows {
+            let def_id = {
+                let def_id = self.id_from_compiler_id(&b.ref_id);
+                if def_id == NULL {
+                    continue;
+                }
+
+                match abs_ref_id(def_id, analysis, project_analysis) {
+                    Some(def_id) => def_id,
+                    None => continue,
+                }
+            };
+
+            let scopes = b.scopes.into_iter().filter_map(|a|
```

Comment:
An easy refactoring would be to factor out lower functions to map rls-data structs to rls-analysis ones. A better, but trickier one would be to make a Lower trait and then use macros 1.1 to derive that for each of the data types.